### PR TITLE
Create preprocess module

### PR DESCRIPTION
**Pull request by roshankern**
@gwaybio This is ready for review!

### CODE REVIEW

**Comment by gwaybio** [comment]
~We just discussed my premature approval. Please warn the user for an OutOfBounds indexing via location col index, and add a test that confirms this warning was thrown~

**Comment by roshankern** [comment]
I think you might have commented on the wrong PR (this is the Cell Health Data PR, not PyCytominer).

**Comment by gwaybio** [comment]
Ha, you're right

File: 3.preprocess-features/3.preprocess-features-env.yml [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+channels:
+  - conda-forge
+dependencies:
+  - conda-forge::python=3.8.13
+  - conda-forge::jupyter=1.0.0
+  - conda-forge::pandas=1.4.2
+  - conda-forge::pip=22.1.2
+  - pip:
+    - pigz-python==1.0.1
+    - git+https://github.com/cytomining/pycytominer
```
Comment (gwaybio) [inline 3.preprocess-features/3.preprocess-features-env.yml]:
```suggestion
    - git+https://github.com/cytomining/pycytominer

```

File: 3.preprocess-features/3.preprocess-features.sh [hunk 2]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+jupyter nbconvert --to python preprocess-features.ipynb
+python preprocess-features.py
```
Comment (gwaybio) [inline 3.preprocess-features/3.preprocess-features.sh]:
```suggestion
python preprocess-features.py

```

File: 3.preprocess-features/README.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+We use [PyCytominer](https://github.com/cytomining/pycytominer) to compile single-cell features.
+
+We use [sklearn.preprocessing.StandardScaler](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html) to derive a normalizion scaler from all negative control features.
+`StandardScaler()` standardizes features by removing the mean and scaling to unit variance.
+
+[Caicedo et al, 2017](https://www.nature.com/articles/nmeth.4397) explain why the negative control features are a good normalization population for our use case:
+> When choosing the normalizing population, we suggest the use of control samples (assuming that they are present in sufficient quantity), because the presence of dramatic phenotypes may confound results. This procedure is good practice regardless of the normalization being performed within plates or across the screen.
+
+The `get_negative_control_index_df` function inside [preprocess-features-utils.py](preprocess-features-utils.py) creates a DeepProfiler-style `index.csv` file (saved in [norm_pop_index.csv](norm_pop_index.csv)) with only the negative control wells.
```
Comment (gwaybio) [inline 3.preprocess-features/README.md]:
```suggestion
The `get_negative_control_index_df` function inside [preprocess-features-utils.py](preprocess-features-utils.py) creates a DeepProfiler-style `index.csv` file (saved in [norm_pop_index.csv](intermediate_files/norm_pop_index.csv)) with only the negative control wells.
```

404 otherwise

File: 3.preprocess-features/README.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+We then use pycytominer to compile the features from these negative control wells and derive and `StandardScaler` with only the negative control features (saved in [negative_control_scaler.save](negative_control_scaler.save)).
+
+After deriving a normalization scaler, we load features by plate and apply the normalization scaler to these plate features.
+The normalized features from each plate are saved in `output_path/`.
+It is necessary to normalize features by plate because we are unable to load the single-cell features for the entire screen into memory.
+
+**Note:** The single-cell dataframes for each plate are first saved as uncompressed files with `pandas.to_csv`.
```
Comment (gwaybio) [inline 3.preprocess-features/README.md]: `pandas.to_csv` can output gzipped files. All you have to do is make the file name end in `.gz` (compression defaults to `infer`) https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_csv.html

File: 3.preprocess-features/README.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+These uncompressed files are then compressed with [pigz-python](https://github.com/nix7drummer88/pigz-python) and the original, uncompressed files are deleted.
+`Pigz-python` is the python implementation of [pigz](https://zlib.net/pigz/), which uses multiple cores to compress files.
+This is significantly faster than the native `pandas` gzip compression.
```
Comment (gwaybio) [inline 3.preprocess-features/README.md]: oh interesting! Let's leave it as is for now then (although I will note that an initial time save might not be better than the risk of a future deprecated python package (`pigz-python` doesn't appear to be widely used))

Comment (gwaybio) [reply 3.preprocess-features/README.md]: How much faster is "significantly"?

File: 3.preprocess-features/README.md [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+`StandardScaler()` standardizes features by removing the mean and scaling to unit variance.
+
+[Caicedo et al, 2017](https://www.nature.com/articles/nmeth.4397) explain why the negative control features are a good normalization population for our use case:
+> When choosing the normalizing population, we suggest the use of control samples (assuming that they are present in sufficient quantity), because the presence of dramatic phenotypes may confound results. This procedure is good practice regardless of the normalization being performed within plates or across the screen.
+
+The `get_negative_control_index_df` function inside [preprocess-features-utils.py](preprocess-features-utils.py) creates a DeepProfiler-style `index.csv` file (saved in [norm_pop_index.csv](norm_pop_index.csv)) with only the negative control wells.
+We then use pycytominer to compile the features from these negative control wells and derive and `StandardScaler` with only the negative control features (saved in [negative_control_scaler.save](negative_control_scaler.save)).
+
+After deriving a normalization scaler, we load features by plate and apply the normalization scaler to these plate features.
```
Comment (gwaybio) [inline 3.preprocess-features/README.md]: I'll take a closer look in the actual code, but I want to confirm that you are creating a normalization scaler _per plate_. A single multi-plate scaler is likely to induce pretty awful batch effects.

File: 3.preprocess-features/preprocess-features-utils.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+        image_annotations = annotations.loc[
+                    (plate == annotations["Plate"]) & (annotations["Well"] == well)
+                ]
+        control_type = image_annotations.iloc[0]["Control Type"]
+        if control_type == "negative":
+            new_index_df.append(row)
+        
+    return pd.DataFrame(new_index_df)
+
+def get_normalization_scaler(norm_pop_index_df_path: pathlib.Path, features_output_dir: pathlib.Path) -> StandardScaler:
```
Comment (gwaybio) [inline 3.preprocess-features/preprocess-features-utils.py:49]: Is this somehow only selecting negative controls? I cannot tell

File: 3.preprocess-features/preprocess-features-utils.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+        metadata = plate_pop[metadata]
+        plate_pop = pd.concat([metadata, features], axis=1)
+        
+        print(f"Saving plate {plate}")
+        # save normalized plate
+        uncompressed_csv_path = pathlib.Path(f"{save_dir}/{plate}_normalized_single_cell.csv")
+        # save csv uncompressed with pandas
+        plate_pop.to_csv(uncompressed_csv_path, index=False)
+        # use pigz to parallel compress csv file
+        pigz_python.compress_file(uncompressed_csv_path)
```
Comment (gwaybio) [inline 3.preprocess-features/preprocess-features-utils.py]: i cringe a bit seeing this - I'm 😨 of python package deprecation! Unless it's supremely faster using pigz_python, I strongly suggest using native pandas compression.

File: 3.preprocess-features/preprocess-features.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+# In[2]:
+
+
+intermediate_files_dir = pathlib.Path("intermediate_files/")
+intermediate_files_dir.mkdir(parents=True, exist_ok=True)
+
+annotations_path = pathlib.Path("../0.image-download/manifest/idr0080-screenA-annotation.csv")
+
+features_output_dir = pathlib.Path("/media/roshankern/63af2010-c376-459e-a56e-576b170133b6/data/cell-health-nuc-DP/outputs/efn_pretrained/features")
+original_index_csv_path = pathlib.Path("/media/roshankern/63af2010-c376-459e-a56e-576b170133b6/data/cell-health-nuc-DP/inputs/metadata/index.csv")
```
Comment (gwaybio) [inline 3.preprocess-features/preprocess-features.py]: remember [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself).

i.e.

```python
base_dir = pathlib.Path("/media/roshankern/63af2010-c376-459e-a56e-576b170133b6/data/cell-health-nuc-DP/")

features_output_dir = pathlib.Path(f"{base_dir}/outputs/efn_pretrained/features")
original_index_csv_path = pathlib.Path(f"{base_dir}/inputs/metadata/index.csv")
```

File: 3.preprocess-features/preprocess-features.py [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+norm_index_df = preprocessUtils.get_negative_control_index_df(index_csv_path, annotations_path)
+norm_index_df.to_csv(normalization_index_csv, index=False)
+
+
+# ### Create/save normalization scaler
+
+# In[5]:
+
+
+scaler_path = f"{intermediate_files_dir}/negative_control_scaler.sclr"
```
Comment (gwaybio) [inline 3.preprocess-features/preprocess-features.py]: Please create per-plate scalers

File: 3.preprocess-features/preprocess-features.py [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+
+scaler = joblib.load(scaler_path)
+
+
+# ### Normalize and save Cell Health single cells by plate
+
+# In[6]:
+
+
+output_path = pathlib.Path("/media/roshankern/63af2010-c376-459e-a56e-576b170133b6/data/cell-health-nuc-normalized/")
```
Comment (gwaybio) [inline 3.preprocess-features/preprocess-features.py]: please define `output_path` in the same cell that you define the other path variables (and remember DRY :) )

**Review by gwaybio — CHANGES_REQUESTED** [review]
Really elegant code, well done.

The main fix is to create per-plate standardscalers. Having only 1 standardscaler is going to induce batch effects 😬

File: 3.preprocess-features/README.md (hunk 5 above)
Comment (roshankern) [reply 3.preprocess-features/README.md]: Can't remember the exact timings for each method, but after discussing with Jenna we have decided to stick with the native compression with `pandas` and not use `pigz-python`.

We agree there is a concern with using a package that might become deprecated in the future.

File: 3.preprocess-features/preprocess-features-utils.py (hunk 7 above)
Comment (roshankern) [reply 3.preprocess-features/preprocess-features-utils.py:49]: The `pycytominer get_single_cells()` function compiles a pandas dataframe from an `index.csv` path. This function derives a scaler from whatever population is specified with the dataframe at `norm_pop_index_df_path`. Thus, when we pass a index df with only the negative controls, this function only selects negative controls to derive the normalization scaler from. Will make this more clear in the documentation

File: 3.preprocess-features/preprocess-features-utils.py (hunk 8 above)
Comment (roshankern) [reply 3.preprocess-features/preprocess-features-utils.py]: Will convert to native pandas compression :ok_hand:

File: 3.preprocess-features/preprocess-features.py (hunk 9 above)
Comment (roshankern) [reply 3.preprocess-features/preprocess-features.py]: Will do!

File: 3.preprocess-features/preprocess-features.py (hunk 11 above)
Comment (roshankern) [reply 3.preprocess-features/preprocess-features.py]: Moved `output_path` in https://github.com/WayScience/cell-health-data/pull/7/commits/c6a2633339debbd3f53ac7ed61375b91b2291ca7 but Jenna and I think it is not a good idea to repeat the hard drive path with DRY for the output path. Instead, we have the user set the path of the DP project to load data from and set `output_path` separately. This way, it is clear that the features can be saved to a completely different path (`output_path`) than they are loaded from (`DP_project_path`).

File: 3.preprocess-features/preprocess-features.py (hunk 10 above)
Comment (roshankern) [reply 3.preprocess-features/preprocess-features.py]: Done in https://github.com/WayScience/cell-health-data/pull/7/commits/a29ab688ca341839c5c6b9e4cf94dc6e3e78914b

File: 3.preprocess-features/README.md (hunk 6 above)
Comment (roshankern) [reply 3.preprocess-features/README.md]: We were not doing normalization per plate but we now are!

File: 3.preprocess-features/preprocess-features-utils.py [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+    """
+    normalize features of DeepProfiler project by plate and save these normalized features in save_dir
+
+    Parameters
+    ----------
+    index_df_path : pathlib.Path
+        path to index df used in DeepProfiler project
+    scaler : StandardScaler
+        scaler to use for normalization
+    features_output_dir : pathlib.Path
```
Comment (gwaybio) [inline 3.preprocess-features/preprocess-features-utils.py:91]: Can you add `plate` as a parameter to this docstring

File: 3.preprocess-features/preprocess-features.py [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+
+# ### Find negative controls for each plate, derive scalers from these negative controls, normalize entire plate with scaler, save normalized plate
+
+# In[4]:
+
+
+# get list of unique plates
+plates = pd.read_csv(index_csv_path)["Metadata_Plate"].unique().tolist()
+
+for plate in plates:
```
Comment (gwaybio) [inline 3.preprocess-features/preprocess-features.py:58]: no content changes, but I do suggest adding some white space (empty lines).

This is really dense code, white space will enable someone to see what's going on with higher clarity.

File: 3.preprocess-features/preprocess-features.py [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+    norm_index_df.to_csv(plate_normalization_index_csv, index=False)
+    # create per-plate normalization scaler from the normalization population
+    print(f"Deriving scaler for plate {plate}")
+    scaler = preprocessUtils.get_normalization_scaler(plate_normalization_index_csv, features_output_dir)
+    # get compiled normalized plate features
+    plate_pop = preprocessUtils.normalize_plate(index_csv_path, scaler, features_output_dir, plate)
+    # save compiled normalized plate features
+    print(f"Saving plate {plate}")
+    normalized_features_csv_path = pathlib.Path(f"{output_path}/{plate}_normalized_single_cell.csv.gz")
+    plate_pop.to_csv(normalized_features_csv_path, index=False, compression='gzip')
```
Comment (gwaybio) [inline 3.preprocess-features/preprocess-features.py:76]: in the notebook, it only looks like 1 plate (`SQ00014617`) processed?

**Review by gwaybio — APPROVED** [review]
LGTM

File: 3.preprocess-features/preprocess-features.py (hunk 14 above)
Comment (roshankern) [reply 3.preprocess-features/preprocess-features.py:76]: We wanted to run the notebook until the last cell so it converted properly to a python file. However, we didn't want to run the entire program as a jupyter notebook.

File: 3.preprocess-features/preprocess-features.py (hunk 13 above)
Comment (roshankern) [reply 3.preprocess-features/preprocess-features.py:58]: Done in https://github.com/WayScience/cell-health-data/pull/7/commits/f2881f98973b71dc4d4e255c378008e6846d1e1b

File: 3.preprocess-features/preprocess-features-utils.py (hunk 12 above)
Comment (roshankern) [reply 3.preprocess-features/preprocess-features-utils.py:91]: Done in https://github.com/WayScience/cell-health-data/pull/7/commits/f2881f98973b71dc4d4e255c378008e6846d1e1b

File: 3.preprocess-features/preprocess-features.py (hunk 14 above)
Comment (roshankern) [reply 3.preprocess-features/preprocess-features.py:76]: Added a comment to explain this in https://github.com/WayScience/cell-health-data/pull/7/commits/dbd53be527badfb34810ab97d8a59468fa5556e6
